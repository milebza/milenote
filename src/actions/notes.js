import uuid from 'uuid'
import database from '../firebase/firebase'

// components calls action generator
// action generator returns function
// component dispatches function
// function runs (has the ability to dispatch other actions and do whatever it wants)

// ADD_NOTE
export const addNote = (note) => ({
  type: 'ADD_NOTE',
  note
})

export const startAddNote = (noteData = {}) => {
  // this function only works due to thunk
  // it gets called internally by redux
  return (dispatch) => {
    // default data
    const {
      title = '',
      content = '',
      date = 0
    } = noteData
    const note = {title, content, date}

    return database.ref('notes').push(note).then((ref) => {
      dispatch(addNote({
        id: ref.key,
        ...note
      }))
    })
  }
}

// REMOVE_NOTE
export const removeNote = ({ id } = {}) => ({
  type: 'REMOVE_NOTE',
  id
})

// EDIT_NOTE
export const editNote = (id, updates) => ({
  type: 'EDIT_NOTE',
  id,
  updates
})
