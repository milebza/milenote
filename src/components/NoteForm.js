import React from 'react'
import moment from 'moment'

export default class NoteForm extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      title: props.note ? props.note.title : '',
      content: props.note ? props.note.content : '',
      date: moment(),
      error: ''
    }
  }
  onTitleChange = (e) => {
    const title = e.target.value
    this.setState(() => ({ title }))
  }
  onContentChange = (e) => {
    const content = e.target.value
    this.setState(() => ({ content }))
  }
  onSubmit = (e) => {
    e.preventDefault()

    if (!this.state.title && !this.state.content) {
      this.setState(() => ({ error: 'Type something!' }))
    } else {
      this.setState(() => ({ error: '' }))
      this.props.onSubmit({
        title: this.state.title,
        content: this.state.content,
        date: this.state.date.valueOf()
      })
    }
  }
  render() {
    return (
      <div>
        <form className="form" onSubmit={this.onSubmit}>
          <div className="form__actions">
            <button className="btn-link form__btn"><span className="icon icon-checkmark"></span> Save</button>
            <p className="form__warning">{this.state.error && this.state.error}</p>
          </div>
          <hr className="mobile-show"></hr>
          <div className="form__input-wrapper">
            <input
              className="form__input form__input--title"
              type="text"
              placeholder="Note title"
              value={this.state.title}
              onChange={this.onTitleChange}
            />
            <textarea
              className="form__input"
              autoFocus
              placeholder="Start typing"
              rows="20"
              value={this.state.content}
              onChange={this.onContentChange}
            />
          </div>
        </form>
      </div>
    )
  }
}
