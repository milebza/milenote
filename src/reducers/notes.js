// Notes Reducer
const NotesReducerDefaultState = []

export default (state = NotesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_NOTE':
      return [
        ...state,
        action.note
      ]
    case 'REMOVE_NOTE':
        return state.filter(({ id }) => (id !== action.id))
    case 'EDIT_NOTE':
      return state.map((note) => {
        if (note.id === action.id) {
          return {
            ...note,
            ...action.updates
          }
        }
        return note
      })
    case 'SET_NOTES':
        return action.notes
    default:
      return state
  }
}
