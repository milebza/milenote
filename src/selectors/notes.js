import moment from 'moment'

// Get visible Notes
export default (notes, {text, startDate, endDate}) => {
  return notes.filter((note) => {
    const date = moment(note.date)
    const startDateMatch = startDate ? startDate.isSameOrBefore(date, 'day') : true
    const endDateMatch = endDate ? endDate.isSameOrAfter(date, 'day') : true
    const textMatch = note.title.toLowerCase().includes(text.toLowerCase()) || note.content.toLowerCase().includes(text.toLowerCase())

    return startDateMatch && endDateMatch && textMatch
  }).sort((a, b) => {
    // 1 -> b will go first
    return a.date < b.date ? 1 : -1
  })
}
