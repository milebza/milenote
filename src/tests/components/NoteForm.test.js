import moment from 'moment'
import React from 'react'
import { shallow } from 'enzyme'
import NoteForm from '../../components/NoteForm'
import notes from '../fixtures/notes'

test('should render NoteForm correctly', () => {
  const wrapper = shallow(<NoteForm />)
  expect(wrapper).toMatchSnapshot()
})

test('should render NoteForm with note data', () => {
  const wrapper = shallow(<NoteForm note={notes[0]} />)
  expect(wrapper).toMatchSnapshot()
})

test('should render error for invalid form submission', () => {
  const wrapper = shallow(<NoteForm />)
  expect(wrapper).toMatchSnapshot()
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  })
  expect(wrapper.state('error').length).toBeGreaterThan(0)
  expect(wrapper).toMatchSnapshot()
})

test('should set title on input change', () => {
  const value = 'New title'
  const wrapper = shallow(<NoteForm />)
  wrapper.find('input').simulate('change', {
    target: { value }
  })
  expect(wrapper.state('title')).toBe(value)
})

test('should set content on textarea change', () => {
  const value = 'New content'
  const wrapper = shallow(<NoteForm />)
  wrapper.find('textarea').simulate('change', {
    target: { value }
  })
  expect(wrapper.state('content')).toBe(value)
})

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn()
  const wrapper = shallow(<NoteForm note={notes[0]} onSubmit={onSubmitSpy} />)
  wrapper.find('form').simulate('submit', {
    preventDefault: () => { }
  })
  expect(wrapper.state('error')).toBe('')
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    title: notes[0].title,
    content: notes[0].content,
    date: expect.any(Number)
  })
})
