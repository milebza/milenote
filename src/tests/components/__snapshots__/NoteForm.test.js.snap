// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render NoteForm correctly 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <button
      className="btn-link form__btn"
    >
      <span
        className="icon icon-checkmark"
      />
       Save
    </button>
    <p
      className="form__warning"
    />
    <div
      className="form__input-wrapper"
    >
      <input
        className="form__input form__input--title"
        onChange={[Function]}
        placeholder="Note title"
        type="text"
        value=""
      />
      <textarea
        autoFocus={true}
        className="form__input"
        onChange={[Function]}
        placeholder="Start typing"
        rows="20"
        value=""
      />
    </div>
  </form>
</div>
`;

exports[`should render NoteForm with note data 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <button
      className="btn-link form__btn"
    >
      <span
        className="icon icon-checkmark"
      />
       Save
    </button>
    <p
      className="form__warning"
    />
    <div
      className="form__input-wrapper"
    >
      <input
        className="form__input form__input--title"
        onChange={[Function]}
        placeholder="Note title"
        type="text"
        value="Letter"
      />
      <textarea
        autoFocus={true}
        className="form__input"
        onChange={[Function]}
        placeholder="Start typing"
        rows="20"
        value="Dear friend"
      />
    </div>
  </form>
</div>
`;

exports[`should render error for invalid form submission 1`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <button
      className="btn-link form__btn"
    >
      <span
        className="icon icon-checkmark"
      />
       Save
    </button>
    <p
      className="form__warning"
    />
    <div
      className="form__input-wrapper"
    >
      <input
        className="form__input form__input--title"
        onChange={[Function]}
        placeholder="Note title"
        type="text"
        value=""
      />
      <textarea
        autoFocus={true}
        className="form__input"
        onChange={[Function]}
        placeholder="Start typing"
        rows="20"
        value=""
      />
    </div>
  </form>
</div>
`;

exports[`should render error for invalid form submission 2`] = `
<div>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <button
      className="btn-link form__btn"
    >
      <span
        className="icon icon-checkmark"
      />
       Save
    </button>
    <p
      className="form__warning"
    >
      Type something!
    </p>
    <div
      className="form__input-wrapper"
    >
      <input
        className="form__input form__input--title"
        onChange={[Function]}
        placeholder="Note title"
        type="text"
        value=""
      />
      <textarea
        autoFocus={true}
        className="form__input"
        onChange={[Function]}
        placeholder="Start typing"
        rows="20"
        value=""
      />
    </div>
  </form>
</div>
`;
